pkg/receive/handler.go

```diff
@@ -533,28 +533,39 @@ func (h *Handler) replicate(ctx context.Context, tenant string, wreq *prompb.Wri
 	}
 	h.mtx.RUnlock()
 
-	var err error
-	ctx, cancel := context.WithTimeout(ctx, h.options.ForwardTimeout)
-	defer func() {
-		// If there is no error, let forward requests optimistically run until timeout.
+	var (
+		quorum       = h.writeQuorum()
+		fctx, cancel = context.WithTimeout(context.Background(), h.options.ForwardTimeout)
+		errChan      = make(chan error)
+	)
+	go func() {
+		close(errChan)
```

Comment:
Closing before doing anything sounds wrong (: You mean defer here?

Reply:
Also I  think we need to defer from main go routine, otherwise we will deadlock on write to errChan

Reply:
🙈 It's what I meant. 
Yeah, I'm gonna fix it.